boards/stm32: use common place of openocd config files for all boards using stlink

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR moves the st-link related codes in `boards/common/nucleo/Makefile.include` in a common place, e.g `boards/common/stm32-stlink`.
This is because not only Nucleo boards use st-link, but also b-l072z-lrwan1 and b-l475e-iot01a boards. The support for these 2 boards is also updated by this PR.

Maybe the stlink related code could go in another place so suggestions are welcome.
Maybe this common code could be reused by stm32fxdiscovery boards since they also use stlink (but some has different openocd configurations).

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Issues/PRs references

None

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

gebart commented:
I like the idea, better configuration reuse, but the interdependence between the board directories is a bit messy (stm32-stlink depends on files in the nucleo common directory)
Is it possible to place some of the configuration in makefiles/tools/ if it is reusable for all st-link users?

aabadie commented:
> Is it possible to place some of the configuration in makefiles/tools/ if it is reusable for all st-link users?

Not sure st-link allows the use of openocd and this is directly related to boards. Maybe move the common openocd in another directory `boards/common/stm32-openocd` ?

aabadie commented:
> the interdependence between the board directories is a bit messy

Indeed, I missed that part...

haukepetersen commented:
Good cleanup! 

I am thinking though: doesn't it rather make sense to move the makefile out of the board tree and dump/merge all the lines from `boards/common/stm32-stlink/Makefile.include` into `makefiles/tools/openocd-adapters/stlink.inc.mk`? I think this would probably be the cleaner and more versatile approach.

aabadie commented:
Updated this PR by moving the common openocd configuration from the common/nucleo to common/stm32-stlink. I think it's better like this.

aabadie commented:
> I think this would probably be the cleaner and more versatile approach.

Sorry, I missed that comment. I'll give this a try.

haukepetersen commented:
cool. (actually, I also saw this file for the first time today :-) ).

Maybe we can add a `cfg` or `config` sub-dir to `dist/tools/openocd/` where we can put all the openocd `*.cfg* files? This way we can maybe improve reusage of them also for many other platforms?

aabadie commented:
> I also saw this file for the first time today :-)

Same here !

> This way we can maybe improve reusage of them also for many other platforms?

That's what I'm trying to do :)

aabadie commented:
I reworked a bit this PR, including previous comments. Here the actual status:
* move stm32 openocd config files to `makefiles/tools/openocd-adapters/configs`. It will certainly be possible to move other openocd configurations there (for other CPUs) and have a common place for all of them
* kept `boards/common/stm32-stlink` to have a single place with common serial, stlink and includes

haukepetersen commented on boards/common/stm32-stlink/Makefile.include:
```diff
@@ -0,0 +1,14 @@
+export STLINK_VERSION ?= 2-1
+
+# Include openocd adapters for st-link
+include $(RIOTMAKE)/tools/openocd-adapters/stlink.inc.mk
```
there seems to be a include-loop: here we include `stlink.inc.mk` and below `openocd.inc.mk`. But `openocd.inc.mk` does also include `stlink.inc.mk`... So seems to me we can drop this line.

haukepetersen commented on boards/common/stm32-stlink/Makefile.include:
```diff
@@ -0,0 +1,14 @@
+
+# Default is to use ST-link v2-1 adapter (used by all Nucleos)
+export DEBUG_ADAPTER ?= stlink
+export STLINK_VERSION ?= 2-1
```
How about to more this line to the top of `makefiles/tools/openocd-adapters/stlink.inc.mk` and add doc that version 2-1 is the default?

haukepetersen commented on boards/common/stm32-stlink/Makefile.include:
```diff
@@ -0,0 +1,14 @@
+# configure the serial terminal
+PORT_LINUX ?= /dev/ttyACM0
+PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.usbmodem*)))
+include $(RIOTMAKE)/tools/serial.inc.mk
```
I think it is not very clean to put the common serial configuration into a common board folder that is referring to a programmer...

I think I have an alternative idea, will PR something later tonight.

haukepetersen replied:
forgot to mention: examples where this might get confusion are for example the green discovery boards, which have an on-board stlink, but don't supply any serial adapter...

aabadie replied:
ok, I was not aware of that.

> I think I have an alternative idea, will PR something later tonight.

Describe what you have in mind and I'll update this PR

haukepetersen replied:
Alright, so how about #8484 for this part.

kaspar030 commented on makefiles/tools/openocd-adapters/stlink.inc.mk:
```diff
@@ -8,3 +8,26 @@ ifneq (,$(DEBUG_ADAPTER_ID))
+
+# select OpenOCD configuration depending on CPU type
+ifeq (,$(OPENOCD_CONFIG))
+  ifeq ($(CPU),stm32f0)
```
Why not ```OPENOCD_CONFIG := $(RIOTMAKE)/tools/openocd-adapters/configs/$(CPU).cfg```?

aabadie commented:
> I am also still convinced, that we can get rid of boards/common/stm32-stlink/Makefile.include

Done, since #7695 might be a good solution for common serial port configuration

> I'd prefer them in boards/common, or maybe in dist, why not just dist/tools/openocd/configs?

I can move them to boards/common, where I think it's simpler to find. In fact, I don't have a strong opinion on that.

aabadie commented:
> Why not OPENOCD_CONFIG := $(RIOTMAKE)/tools/openocd-adapters/configs/$(CPU).cfg?

Good point! Completely missed that factorisation. Done !

aabadie commented:
> I'd prefer them in boards/common, or maybe in dist, why not just dist/tools/openocd/configs?

Thinking of it now and it makes more sense to put the common openocd config files in `boards/common`. This is because those files contains openocd options that are specific to some boards.
Why not `boards/common/stm32/openocd` ?

kYc0o commented:
> Thinking of it now and it makes more sense to put the common openocd config files in boards/common. This is because those files contains openocd options that are specific to some boards.
Why not boards/common/stm32/openocd ?

I don't think it's directly bound to stm32, since that's a cpu and the config depends on the board. For this, I would say a bigger refactor on the boards folder is necessary, on which we can maybe separate the boards according to it's CPUs, or a similar approach.

However, it's also not bound to nucleo devices, even though they're very similar.

I'd vote for something like boards/common/st, which is bound only to the manufacturer.

aabadie commented:
> I'd vote for something like boards/common/st, which is bound only to the manufacturer.

+1

ping @kaspar030 @haukepetersen, is it ok for you too ?

aabadie commented:
> ping @kaspar030 @haukepetersen, is it ok for you too ?

ping bis

aabadie commented:
@kYc0o, I readapted like you suggested and also applied the modification to the stm32fxdiscovery boards.

aabadie commented:
FYI, I tested the stm32f4discovery and stm32f3discovery boards with success.

aabadie commented:
@gebart, @kYc0o are you ok with this one ? @haukepetersen, @kaspar030 do you have other comments ?

kYc0o commented on boards/stm32f4discovery/Makefile.include:
```diff
@@ -11,6 +11,7 @@ include $(RIOTMAKE)/tools/serial.inc.mk
 
 export DEBUG_ADAPTER ?= stlink
 export STLINK_VERSION ?= 2
```
Why here the version is still important and for the others you have moved it?

aabadie replied:
discovery boards use stlink v2 and nucleo boards use v2-1. The default in the stlink.mk.inc file is 2-1, that's why it has to be specified here.

dylad replied:
F4 doesn't work with ST-LINK 2-1 AFAIK so we must keep this version here.

aabadie replied:
>  for f7 you removed it

Because it was already using 2-1. I left the ones already using v2

aabadie replied:
To summarize the discovery changes:
- f0 and f4: use stlink v2 but a different openocd script => force v2 and the board specific openocd script
- f3: use stlink v2 and the same openocd script as nucleo f3 boards  => force v2 and remove the openocd script so that the common one (in common/st/dist) will be used
- f7: use stlink v2-1 and (nearly) the same openocd script => remove redefinition of v2-1 (useless) and remove the openocd script. Maybe I can keep the previous openocd script but I don't think it hurts the way I did.

kYc0o replied:
Ok, I understand now. Thanks for the explanation!

aabadie commented:
@cladmi, I addressed all your comments from #8976 here. Maybe you can re-review this one ?

aabadie commented:
@kaspar030, @haukepetersen, @cladmi are we good here ?

aabadie commented:
ping @kaspar030, @haukepetersen, @cladmi

aabadie commented:
Any chance to have a final review here @kaspar030, @haukepetersen ? It's blocking other related PRs (#8868, #9135, etc). Thanks !

cladmi commented:
Will do in Hack'n'Hack.

cladmi commented on makefiles/tools/openocd-adapters/stlink.inc.mk:
```diff
@@ -8,3 +11,15 @@ ifneq (,$(DEBUG_ADAPTER_ID))
+# select OpenOCD configuration if not already set by the board configuration
+ifeq (,$(OPENOCD_CONFIG))
+  OPENOCD_BOARD_CFG = $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg
+  ifeq "$(OPENOCD_BOARD_CFG)" "$(wildcard $(OPENOCD_BOARD_CFG))"
```
You do not need to set `OPENOCD_CONFIG` when it is the default value as it will be set in `openocd.sh`

So something like this would work I think.

``` diff
diff --git a/makefiles/tools/openocd-adapters/stlink.inc.mk b/makefiles/tools/openocd-adapters/stlink.inc.mk
index 6e7fdc6..0a5e008 100644
--- a/makefiles/tools/openocd-adapters/stlink.inc.mk
+++ b/makefiles/tools/openocd-adapters/stlink.inc.mk
@@ -14,11 +14,8 @@ export OPENOCD_ADAPTER_INIT
 
 # select OpenOCD configuration if not already set by the board configuration
 ifeq (,$(OPENOCD_CONFIG))
-  OPENOCD_BOARD_CFG = $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg
-  ifeq "$(OPENOCD_BOARD_CFG)" "$(wildcard $(OPENOCD_BOARD_CFG))"
-    # Use the openocd config file in the board directory if it exists
-    export OPENOCD_CONFIG := $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg
-  else
+  # select OpenOCD configuration if there is none in the default location
+  ifeq (,$(wildcard $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg))
     # select OpenOCD configuration depending on CPU type
     export OPENOCD_CONFIG := $(RIOTBASE)/boards/common/st/configs/$(CPU).cfg
   endif
```

I checked with this both in master and this pr with the diff.
And checked that for `stm32f769i-disco` it is the default file in `openocd.sh`.
``` bash
$ make --no-print-directory -C examples/hello-world/ BOARD=stm32f769i-disco info-debug-variable-OPENOCD_CONFIG

$ make --no-print-directory -C examples/hello-world/ BOARD=nucleo-l152re info-debug-variable-OPENOCD_CONFIG
/home/harter/work/git/RIOT/boards/common/st/configs/stm32l1.cfg
```

aabadie replied:
This is what I did initially but for some reason it was not working: the `ifeq` was always false and the common configuration always used. I tested that with a nucleo-l073rz and b-l072z-blabla

cladmi replied:
It works on my computer.

With the diff:

```
make --no-print-directory -C examples/hello-world/ BOARD=nucleo-l073rz info-debug-variable-OPENOCD_CONFIG
/home/harter/work/git/RIOT/boards/common/st/configs/stm32l0.cfg
```

For b-l072-lrwan-1 it is using the one in dist/openocd.cfg
```
make --no-print-directory -C examples/hello-world/ BOARD=b-l072z-lrwan1 info-debug-variable-OPENOCD_CONFIG

# Which is wanted as it has a different configuration
diff -u boards/b-l072z-lrwan1/dist/openocd.cfg boards/common/st/configs/stm32l0.cfg
--- boards/b-l072z-lrwan1/dist/openocd.cfg	2018-05-14 15:15:07.324865950 +0200
+++ boards/common/st/configs/stm32l0.cfg	2018-05-29 22:49:54.915624904 +0200
@@ -1,3 +1,3 @@
 source [find target/stm32l0.cfg]
-
-reset_config trst_and_srst srst_nogate connect_assert_srst
+reset_config srst_only
+$_TARGETNAME configure -rtos auto
```

cladmi replied:
If there really is a whitespace non-robustness problem it can be done with `words` also

`ifeq (0,$(words $(wildcard $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg)))`

aabadie replied:
Using make words function does the trick.

aabadie commented:
> The stm32f3discovery modification are currently in the stm32f76i-disco commit which looks odd.

I can split 9385222a5f8f05e456d1702bc9319a787719cafb in 2 if you prefer

aabadie commented on makefiles/tools/openocd-adapters/stlink.inc.mk:
```diff
@@ -8,3 +11,14 @@ ifneq (,$(DEBUG_ADAPTER_ID))
+ifeq (,$(OPENOCD_CONFIG))
+  ifneq (0,$(words $(wildcard $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg)))
+    # Use the openocd config file in the board directory if it exists
+    export OPENOCD_CONFIG := $(RIOTBOARD)/$(BOARD)/dist/openocd.cfg
```
You are right! I updated the file with a detailed comment.

aabadie commented:
@cladmi, I update the PR. Hope this is fine now.

cladmi commented on boards/b-l072z-lrwan1/dist/openocd.cfg:
```diff
@@ -1,3 +1,3 @@
 source [find target/stm32l0.cfg]
 
-reset_config trst_and_srst srst_nogate connect_assert_srst
+reset_config srst_only connect_assert_srst
```
Change not related.

aabadie replied:
Ok, I dropped the commit, it was the same change as #8790, so can be reviewed and merged from there.

aabadie commented:
squahed. @haukepetersen @kaspar030 do you still have comments or can we dismiss your reviews since I think your previous comments were addressed ?

aabadie commented:
@cladmi @kaspar030 @haukepetersen, now that #8549 is merged, putting the openocd common configuration file in `boards/stm32` makes more sense to me. What do you think ?

kaspar030 commented:
This looks OK to me.

aabadie commented:
@cladmi, @haukepetersen do you have more comments here ? or can I squash ?

aabadie commented:
@haukepetersen @cladmi, ok to squash here ?

cladmi commented:
Ok for me

aabadie commented:
It would be great if someone could ping @haukepetersen IRL or at least dismiss his review.

aabadie commented:
Seems that @haukepetersen will never reply here. @cladmi @kaspar030, are you ok for dismissing his review ?

aabadie commented:
@haukepetersen, if you don't reply or if you are not interested anymore by this PR, please let us know.

aabadie commented:
rebased and squashed. @cladmi, I think we left enough time to @haukepetersen to leave his opinion her and since it never came, I'll dismiss his review.

aabadie commented:
Let's merge this one.
